### ISSUE TITLE
luminous: osd auth add fails when deploying on kubernetes

### ISSUE THREAD

**Issue by r0ss3** [ceph/ceph-container#820]
I'm testing with luminous on a kubernetes cluster.  The mons are ok, but none of the osd's are able to start.  They get a "Error EACCES: access denied" when when doing the ceph auth add.

As far as I can see the bootstrap keyring is ok:

```
[root@ceph-osd-7sgr6 /]# cat /var/lib/ceph/bootstrap-osd/ceph-test.keyring
[client.bootstrap-osd]
  key = AQDlat9ZAAAAABAAXLQPDaLJuyObDog1H41r3Q==
  caps mon = "allow profile bootstrap-osd"
```

key matches the one in the mons, and the profile is bootstrap-osd
```
[root@ceph-mon-2339881336-g447h /]# ceph auth list --cluster=ceph-test
installed auth entries:
...
client.bootstrap-osd
        key: AQDlat9ZAAAAABAAXLQPDaLJuyObDog1H41r3Q==
        caps: [mon] allow profile bootstrap-osd
...
```

Full debug log:

```
2017-10-12 13:15:35  /entrypoint.sh: VERBOSE: activating bash debugging mode.
2017-10-12 13:15:35  /entrypoint.sh: To run Ceph daemons in debugging mode, pass the CEPH_ARGS variable like this:
2017-10-12 13:15:35  /entrypoint.sh: -e CEPH_ARGS='--debug-ms 1 --debug-osd 10'
++debug.sh:10: for option in '$(comma_to_space "${DEBUG}")'
++debug.sh:11: case $option in
++debug.sh:57: log 'STAYALIVE: container will not die if a command fails.'
++common_functions.sh:6: log(): '[' -z 'STAYALIVE: container will not die if a command fails.' ']'
++common_functions.sh:10: log(): local timestamp
2017-10-12 13:15:35  /entrypoint.sh: This container environement variables are: CEPH_MGR_PORT_80_TCP=tcp://172.22.162.156:80
HOSTNAME=ceph-osd-7sgr6
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT=tcp://172.20.0.1:443
KUBERNETES_SERVICE_PORT=443
CEPH_MGR_PORT_80_TCP_PORT=80
KUBERNETES_SERVICE_HOST=172.20.0.1
CEPH_MGR_PORT_80_TCP_PROTO=tcp
LC_ALL=C
CEPH_MGR_PORT=tcp://172.22.162.156:80
CEPH_MGR_SERVICE_PORT=80
CEPH_GET_ADMIN_KEY=1
CEPH_MGR_PORT_80_TCP_ADDR=172.22.162.156
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
CEPH_VERSION=luminous
SHLVL=1
HOME=/root
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_SERVICE_PORT_HTTPS=443
CEPH_MGR_SERVICE_HOST=172.22.162.156
CLUSTER=ceph-test
CONFD_VERSION=0.10.0
KUBERNETES_PORT_443_TCP_ADDR=172.20.0.1
KV_TYPE=k8s
CEPH_DAEMON=osd_directory
KUBERNETES_PORT_443_TCP=tcp://172.20.0.1:443
DEBUG=verbose,stayalive
_=/usr/bin/env
+++common_functions.sh:11: log(): date '+%F %T'
2017-10-12 13:15:35  /entrypoint.sh: STAYALIVE: container will not die if a command fails.
++common_functions.sh:11: log(): timestamp='2017-10-12 13:15:35'
++common_functions.sh:12: log(): echo '2017-10-12 13:15:35  /entrypoint.sh: STAYALIVE: container will not die if a command fails.'
++common_functions.sh:13: log(): return 0
++debug.sh:58: source docker_exec.sh
+++docker_exec.sh:11: NOTRAP=
+++docker_exec.sh:33: child_for_exec=1
+++docker_exec.sh:65: set_trap_err
+++docker_exec.sh:5: set_trap_err(): set -E
+++docker_exec.sh:8: set_trap_err(): trap trap_error ERR
+/entrypoint.sh:15: case "$KV_TYPE" in
+/entrypoint.sh:22: source /config.k8s.sh
++/config.k8s.sh:2: set -e
++/entrypoint.sh:35: to_lowercase osd_directory
++common_functions.sh:177: to_lowercase(): echo osd_directory
+/entrypoint.sh:35: CEPH_DAEMON=osd_directory
+/entrypoint.sh:37: create_mandatory_directories
+common_functions.sh:52: create_mandatory_directories(): for keyring in '$OSD_BOOTSTRAP_KEYRING' '$MDS_BOOTSTRAP_KEYRING' '$RGW_BOOTSTRAP_KEYRING' '$RBD_MIRROR_BOOTSTRAP_KEYRING'
++common_functions.sh:53: create_mandatory_directories(): dirname /var/lib/ceph/bootstrap-osd/ceph-test.keyring
+common_functions.sh:53: create_mandatory_directories(): mkdir -p /var/lib/ceph/bootstrap-osd
+common_functions.sh:52: create_mandatory_directories(): for keyring in '$OSD_BOOTSTRAP_KEYRING' '$MDS_BOOTSTRAP_KEYRING' '$RGW_BOOTSTRAP_KEYRING' '$RBD_MIRROR_BOOTSTRAP_KEYRING'
++common_functions.sh:53: create_mandatory_directories(): dirname /var/lib/ceph/bootstrap-mds/ceph-test.keyring
+common_functions.sh:53: create_mandatory_directories(): mkdir -p /var/lib/ceph/bootstrap-mds
+common_functions.sh:52: create_mandatory_directories(): for keyring in '$OSD_BOOTSTRAP_KEYRING' '$MDS_BOOTSTRAP_KEYRING' '$RGW_BOOTSTRAP_KEYRING' '$RBD_MIRROR_BOOTSTRAP_KEYRING'
++common_functions.sh:53: create_mandatory_directories(): dirname /var/lib/ceph/bootstrap-rgw/ceph-test.keyring
+common_functions.sh:53: create_mandatory_directories(): mkdir -p /var/lib/ceph/bootstrap-rgw
+common_functions.sh:52: create_mandatory_directories(): for keyring in '$OSD_BOOTSTRAP_KEYRING' '$MDS_BOOTSTRAP_KEYRING' '$RGW_BOOTSTRAP_KEYRING' '$RBD_MIRROR_BOOTSTRAP_KEYRING'
++common_functions.sh:53: create_mandatory_directories(): dirname /var/lib/ceph/bootstrap-rbd/ceph-test.keyring
+common_functions.sh:53: create_mandatory_directories(): mkdir -p /var/lib/ceph/bootstrap-rbd
+common_functions.sh:57: create_mandatory_directories(): for directory in mon osd mds radosgw tmp mgr
+common_functions.sh:58: create_mandatory_directories(): mkdir -p /var/lib/ceph/mon
+common_functions.sh:57: create_mandatory_directories(): for directory in mon osd mds radosgw tmp mgr
+common_functions.sh:58: create_mandatory_directories(): mkdir -p /var/lib/ceph/osd
+common_functions.sh:57: create_mandatory_directories(): for directory in mon osd mds radosgw tmp mgr
+common_functions.sh:58: create_mandatory_directories(): mkdir -p /var/lib/ceph/mds
+common_functions.sh:57: create_mandatory_directories(): for directory in mon osd mds radosgw tmp mgr
+common_functions.sh:58: create_mandatory_directories(): mkdir -p /var/lib/ceph/radosgw
+common_functions.sh:57: create_mandatory_directories(): for directory in mon osd mds radosgw tmp mgr
+common_functions.sh:58: create_mandatory_directories(): mkdir -p /var/lib/ceph/tmp
+common_functions.sh:57: create_mandatory_directories(): for directory in mon osd mds radosgw tmp mgr
+common_functions.sh:58: create_mandatory_directories(): mkdir -p /var/lib/ceph/mgr
+common_functions.sh:62: create_mandatory_directories(): mkdir -p /var/lib/ceph/mon/ceph-test-ceph-osd-7sgr6
+common_functions.sh:65: create_mandatory_directories(): mkdir -p /var/run/ceph
+common_functions.sh:68: create_mandatory_directories(): mkdir -p /var/lib/ceph/radosgw/ceph-test-rgw.ceph-osd-7sgr6
+common_functions.sh:71: create_mandatory_directories(): mkdir -p /var/lib/ceph/mds/ceph-test-ceph-osd-7sgr6
+common_functions.sh:74: create_mandatory_directories(): mkdir -p /var/lib/ceph/mgr/ceph-test-ceph-osd-7sgr6
+common_functions.sh:77: create_mandatory_directories(): chown --verbose -R ceph. /var/run/ceph/
ownership of '/var/run/ceph/' retained as ceph:ceph
+common_functions.sh:78: create_mandatory_directories(): find -L /var/lib/ceph/ -mindepth 1 -maxdepth 3 -exec chown --verbose ceph. '{}' ';'
changed ownership of '/var/lib/ceph/bootstrap-rbd' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-rbd/..data' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-rbd/..data/ceph-test.keyring' from root:root to ceph:ceph
ownership of '/var/lib/ceph/bootstrap-rbd/ceph-test.keyring' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-rbd/..109810_12_10_15_15_20.682555425' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-rbd/..109810_12_10_15_15_20.682555425/ceph-test.keyring' retained as ceph:ceph
changed ownership of '/var/lib/ceph/osd' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-rgw' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-rgw/..data' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-rgw/..data/ceph-test.keyring' from root:root to ceph:ceph
ownership of '/var/lib/ceph/bootstrap-rgw/ceph-test.keyring' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-rgw/..109810_12_10_15_15_20.641370200' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-rgw/..109810_12_10_15_15_20.641370200/ceph-test.keyring' retained as ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-mds' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-mds/..data' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-mds/..data/ceph-test.keyring' from root:root to ceph:ceph
ownership of '/var/lib/ceph/bootstrap-mds/ceph-test.keyring' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-mds/..109810_12_10_15_15_20.211353930' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-mds/..109810_12_10_15_15_20.211353930/ceph-test.keyring' retained as ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-osd' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-osd/..data' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/bootstrap-osd/..data/ceph-test.keyring' from root:root to ceph:ceph
ownership of '/var/lib/ceph/bootstrap-osd/ceph-test.keyring' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-osd/..109810_12_10_15_15_20.495096279' retained as ceph:ceph
ownership of '/var/lib/ceph/bootstrap-osd/..109810_12_10_15_15_20.495096279/ceph-test.keyring' retained as ceph:ceph
changed ownership of '/var/lib/ceph/mon' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/mon/ceph-test-ceph-osd-7sgr6' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/mds' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/mds/ceph-test-ceph-osd-7sgr6' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/radosgw' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/radosgw/ceph-test-rgw.ceph-osd-7sgr6' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/tmp' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/mgr' from root:root to ceph:ceph
changed ownership of '/var/lib/ceph/mgr/ceph-test-ceph-osd-7sgr6' from root:root to ceph:ceph
+/entrypoint.sh:41: case "$CEPH_DAEMON" in
+/entrypoint.sh:59: source start_osd.sh
++start_osd.sh:2: set -e
++start_osd.sh:4: is_redhat
++common_functions.sh:210: is_redhat(): get_package_manager
++common_functions.sh:195: get_package_manager(): is_available rpm
++common_functions.sh:46: is_available(): command -v rpm
++common_functions.sh:196: get_package_manager(): OS_VENDOR=redhat
++common_functions.sh:211: is_redhat(): [[ redhat == \r\e\d\h\a\t ]]
++start_osd.sh:5: source /etc/sysconfig/ceph
+++/etc/sysconfig/ceph:7: TCMALLOC_MAX_TOTAL_THREAD_CACHE_BYTES=134217728
+++/etc/sysconfig/ceph:26: CEPH_AUTO_RESTART_ON_UPGRADE=no
+/entrypoint.sh:60: OSD_TYPE=directory
+/entrypoint.sh:61: start_osd
+start_osd.sh:11: start_osd(): get_config
+/config.k8s.sh:41: get_config(): log 'k8s: config is stored as k8s secrets.'
+common_functions.sh:6: log(): '[' -z 'k8s: config is stored as k8s secrets.' ']'
+common_functions.sh:10: log(): local timestamp
++common_functions.sh:11: log(): date '+%F %T'
2017-10-12 13:15:35  /entrypoint.sh: k8s: config is stored as k8s secrets.
+common_functions.sh:11: log(): timestamp='2017-10-12 13:15:35'
+common_functions.sh:12: log(): echo '2017-10-12 13:15:35  /entrypoint.sh: k8s: config is stored as k8s secrets.'
+common_functions.sh:13: log(): return 0
+start_osd.sh:12: start_osd(): check_config
+common_functions.sh:18: check_config(): [[ ! -e /etc/ceph/ceph-test.conf ]]
+start_osd.sh:14: start_osd(): '[' 1 -eq 1 ']'
+start_osd.sh:15: start_osd(): get_admin_key
+/config.k8s.sh:6: get_admin_key(): log 'k8s: does not generate the admin key. Use Kubernetes secrets instead.'
+common_functions.sh:6: log(): '[' -z 'k8s: does not generate the admin key. Use Kubernetes secrets instead.' ']'
+common_functions.sh:10: log(): local timestamp
++common_functions.sh:11: log(): date '+%F %T'
2017-10-12 13:15:35  /entrypoint.sh: k8s: does not generate the admin key. Use Kubernetes secrets instead.
+common_functions.sh:11: log(): timestamp='2017-10-12 13:15:35'
+common_functions.sh:12: log(): echo '2017-10-12 13:15:35  /entrypoint.sh: k8s: does not generate the admin key. Use Kubernetes secrets instead.'
+common_functions.sh:13: log(): return 0
+start_osd.sh:16: start_osd(): check_admin_key
+common_functions.sh:26: check_admin_key(): [[ ! -e /etc/ceph/ceph-test.client.admin.keyring ]]
+start_osd.sh:19: start_osd(): case "$OSD_TYPE" in
+start_osd.sh:21: start_osd(): source osd_directory.sh
++osd_directory.sh:2: source(): set -e
+start_osd.sh:22: start_osd(): source osd_common.sh
++osd_common.sh:2: source(): set -e
+start_osd.sh:23: start_osd(): osd_directory
+osd_directory.sh:5: osd_directory(): [[ ! -d /var/lib/ceph/osd ]]
++osd_directory.sh:12: osd_directory(): find /var/lib/ceph/osd -prune -empty
+osd_directory.sh:12: osd_directory(): [[ -n /var/lib/ceph/osd ]]
+osd_directory.sh:13: osd_directory(): log 'Creating osd with ceph --cluster ceph-test osd create'
+common_functions.sh:6: log(): '[' -z 'Creating osd with ceph --cluster ceph-test osd create' ']'
+common_functions.sh:10: log(): local timestamp
++common_functions.sh:11: log(): date '+%F %T'
2017-10-12 13:15:35  /entrypoint.sh: Creating osd with ceph --cluster ceph-test osd create
+common_functions.sh:11: log(): timestamp='2017-10-12 13:15:35'
+common_functions.sh:12: log(): echo '2017-10-12 13:15:35  /entrypoint.sh: Creating osd with ceph --cluster ceph-test osd create'
+common_functions.sh:13: log(): return 0
++osd_directory.sh:14: osd_directory(): ceph --cluster ceph-test osd create
+osd_directory.sh:14: osd_directory(): OSD_ID=2
+osd_directory.sh:15: osd_directory(): is_integer 2
+common_functions.sh:170: is_integer(): for arg in '"$@"'
+common_functions.sh:171: is_integer(): [[ 2 =~ ^-?[0-9]+$ ]]
+osd_directory.sh:16: osd_directory(): log 'OSD created with ID: 2'
+common_functions.sh:6: log(): '[' -z 'OSD created with ID: 2' ']'
+common_functions.sh:10: log(): local timestamp
++common_functions.sh:11: log(): date '+%F %T'
+common_functions.sh:11: log(): timestamp='2017-10-12 13:15:38'
2017-10-12 13:15:38  /entrypoint.sh: OSD created with ID: 2
+common_functions.sh:12: log(): echo '2017-10-12 13:15:38  /entrypoint.sh: OSD created with ID: 2'
+common_functions.sh:13: log(): return 0
++osd_directory.sh:22: osd_directory(): get_osd_path 2
++common_functions.sh:241: get_osd_path(): echo /var/lib/ceph/osd/ceph-test-2/
+osd_directory.sh:22: osd_directory(): OSD_PATH=/var/lib/ceph/osd/ceph-test-2/
+osd_directory.sh:25: osd_directory(): mkdir -p /var/lib/ceph/osd/ceph-test-2/
+osd_directory.sh:26: osd_directory(): chown --verbose ceph. /var/lib/ceph/osd/ceph-test-2/
changed ownership of '/var/lib/ceph/osd/ceph-test-2/' from root:root to ceph:ceph
+osd_directory.sh:27: osd_directory(): log 'created folder /var/lib/ceph/osd/ceph-test-2/'
+common_functions.sh:6: log(): '[' -z 'created folder /var/lib/ceph/osd/ceph-test-2/' ']'
+common_functions.sh:10: log(): local timestamp
++common_functions.sh:11: log(): date '+%F %T'
2017-10-12 13:15:38  /entrypoint.sh: created folder /var/lib/ceph/osd/ceph-test-2/
+common_functions.sh:11: log(): timestamp='2017-10-12 13:15:38'
+common_functions.sh:12: log(): echo '2017-10-12 13:15:38  /entrypoint.sh: created folder /var/lib/ceph/osd/ceph-test-2/'
+common_functions.sh:13: log(): return 0
+osd_directory.sh:31: osd_directory(): mkdir -p /etc/forego/ceph-test
+osd_directory.sh:32: osd_directory(): echo ''
++osd_directory.sh:4: osd_directory(): find /var/lib/ceph/osd -maxdepth 1 -name 'ceph-test*'
++osd_directory.sh:4: osd_directory(): sed 's/.*-//'
+osd_directory.sh:34: osd_directory(): for OSD_ID in '$(find /var/lib/ceph/osd -maxdepth 1 -name "${CLUSTER}*" | sed '\''s/.*-//'\'')'
++osd_directory.sh:35: osd_directory(): get_osd_path 2
++common_functions.sh:241: get_osd_path(): echo /var/lib/ceph/osd/ceph-test-2/
+osd_directory.sh:35: osd_directory(): OSD_PATH=/var/lib/ceph/osd/ceph-test-2/
+osd_directory.sh:36: osd_directory(): OSD_KEYRING=/var/lib/ceph/osd/ceph-test-2//keyring
+osd_directory.sh:38: osd_directory(): '[' -n '' ']'
+osd_directory.sh:42: osd_directory(): '[' -n '' ']'
+osd_directory.sh:46: osd_directory(): OSD_J=/var/lib/ceph/osd/ceph-test-2//journal
+osd_directory.sh:50: osd_directory(): '[' '!' -e /var/lib/ceph/osd/ceph-test-2//keyring ']'
+osd_directory.sh:51: osd_directory(): chown --verbose ceph. /var/lib/ceph/osd/ceph-test-2/
ownership of '/var/lib/ceph/osd/ceph-test-2/' retained as ceph:ceph
+osd_directory.sh:53: osd_directory(): ceph-osd --cluster ceph-test -i 2 --mkfs --mkkey --mkjournal --osd-journal /var/lib/ceph/osd/ceph-test-2//journal --setuser ceph --setgroup ceph
2017-10-12 13:15:39.056265 7f5197bcfd00 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2017-10-12 13:15:39.069716 7f5197bcfd00 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2017-10-12 13:15:39.070498 7f5197bcfd00 -1 read_settings error reading settings: (2) No such file or directory
2017-10-12 13:15:39.082862 7f5197bcfd00 -1 created object store /var/lib/ceph/osd/ceph-test-2 for osd.2 fsid 92a424cf-b34a-4ebc-9b11-93538c958cc8
2017-10-12 13:15:39.082892 7f5197bcfd00 -1 auth: error reading file: /var/lib/ceph/osd/ceph-test-2/keyring: can't open /var/lib/ceph/osd/ceph-test-2/keyring: (2) No such file or directory
2017-10-12 13:15:39.083021 7f5197bcfd00 -1 created new key in keyring /var/lib/ceph/osd/ceph-test-2/keyring
+osd_directory.sh:54: osd_directory(): '[' '!' -e /var/lib/ceph/bootstrap-osd/ceph-test.keyring ']'
+osd_directory.sh:58: osd_directory(): ceph_health client.bootstrap-osd /var/lib/ceph/bootstrap-osd/ceph-test.keyring
+common_functions.sh:300: ceph_health(): local bootstrap_user=client.bootstrap-osd
+common_functions.sh:301: ceph_health(): local bootstrap_key=/var/lib/ceph/bootstrap-osd/ceph-test.keyring
+common_functions.sh:303: ceph_health(): timeout 10 ceph --cluster ceph-test --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph-test.keyring health
HEALTH_OK
+osd_directory.sh:61: osd_directory(): ceph --cluster ceph-test --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph-test.keyring auth add osd.2 -i /var/lib/ceph/osd/ceph-test-2//keyring osd 'allow *' mon 'allow profile osd'
Error EACCES: access denied
+osd_directory.sh:61: osd_directory(): log ''
+common_functions.sh:6: log(): '[' -z '' ']'
+common_functions.sh:7: log(): return 1
++osd_directory.sh:7: osd_directory(): trap_error
++docker_exec.sh:13: trap_error(): set +x
An issue occured and you asked me to stay alive.
You can connect to me with: sudo docker exec -i -t ceph-osd-7sgr6 /bin/bash
The current environment variables will be reloaded by this bash to be in a similar context.
When debugging is over stop me with: pkill sleep
I'll sleep for 365 days waiting for you darling, bye bye
```

Anyone an idea how I can debug this further?

**Comment by danelowe** [comment]
Took me a long time to find, but I believe it may be solved by #819, and is cause of #829

**Comment by r0ss3** [comment]
Good find!  This works now, so I'll close.
